File: download/src/main/java/com/bbou/download/FileDataDownloader.java
/*
 * Copyright (c) 2023. Bernard Bou <1313ou@gmail.com>.
 */

package com.bbou.download;

import android.app.Activity;
import android.content.Intent;
import android.util.Log;
import android.widget.Toast;

import com.bbou.concurrency.Task;

import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLConnection;
import java.util.Date;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * File data downloader
 *
 * @author <a href="mailto:1313ou@gmail.com">Bernard Bou</a>
 */
public class FileDataDownloader extends Task<String, Void, FileData>
{
	static private final String TAG = "FileDataDownloader";

	/**
	 * Result listener
	 */
	private final Listener listener;

	/**
	 * Constructor
	 *
	 * @param listener listener
	 */
	private FileDataDownloader(final Listener listener)
	{
		this.listener = listener;
	}

	@Nullable
	@Override
	protected FileData doInBackground(final String... params)
	{
		final String urlString = params[0];

		long date = -1;
		long size = -1;
		String etag = null;
		String version = null;
		String staticVersion = null;

		HttpURLConnection httpConnection = null;
		try
		{
			// url
			final URL url = new URL(urlString);
			Log.d(TAG, "Getting " + url);

			// connection
			URLConnection connection = url.openConnection();

			// handle redirect
			if (connection instanceof HttpURLConnection)
			{
				httpConnection = (HttpURLConnection) connection;
				httpConnection.setInstanceFollowRedirects(false);
				HttpURLConnection.setFollowRedirects(false);

				int status = httpConnection.getResponseCode();
				Log.d(TAG, "Response Code ... " + status);
				if (status == HttpURLConnection.HTTP_MOVED_TEMP || status == HttpURLConnection.HTTP_MOVED_PERM || status == HttpURLConnection.HTTP_SEE_OTHER)
				{
					// headers
					date = connection.getLastModified(); // new Date(date));
					size = connection.getContentLength();
					etag = connection.getHeaderField("etag");
					version = connection.getHeaderField("x-version");
					staticVersion = connection.getHeaderField("x-static-version");

					// get redirect url from "location" header field
					String newUrl = httpConnection.getHeaderField("Location");

					// close
					httpConnection.getInputStream().close();

					// disconnect
					httpConnection.disconnect();

					// open the new connection again
					connection = httpConnection = (HttpURLConnection) new URL(newUrl).openConnection();
					httpConnection.setInstanceFollowRedirects(true);
					HttpURLConnection.setFollowRedirects(true);
					Log.d(TAG, "Redirect to URL : " + newUrl);
				}
			}

			// connect
			connection.connect();

			// expect HTTP 200 OK, so we don't mistakenly save error report instead of the file
			if (connection instanceof HttpURLConnection)
			{
				if (httpConnection.getResponseCode() != HttpURLConnection.HTTP_OK)
				{
					final String message = "server returned HTTP " + httpConnection.getResponseCode() + " " + httpConnection.getResponseMessage();
					throw new RuntimeException(message);
				}
			}
			final String name = url.getFile();
			if (date <= 0)
			{
				date = connection.getLastModified(); // new Date(date));
			}
			if (size <= 0)
			{
				size = connection.getContentLength();
			}
			if (etag == null)
			{
				etag = connection.getHeaderField("etag");
			}
			if (version == null)
			{
				version = connection.getHeaderField("x-version");
			}
			if (staticVersion == null)
			{
				staticVersion = connection.getHeaderField("x-static-version");
			}

			// close
			connection.getInputStream().close();

			return new FileData(name, date, size, etag, version, staticVersion);
		}
		catch (@NonNull final Exception e)
		{
			Log.e(TAG, "While downloading", e);
		}
		finally
		{
			if (httpConnection != null)
			{
				httpConnection.disconnect();
			}
		}
		return null;
	}

	@Override
	protected void onPostExecute(final FileData result)
	{
		Log.d(TAG, "Completed " + result);
		this.listener.onDone(result);
	}

	@Override
	protected void onCancelled(final FileData result)
	{
		Log.d(TAG, "Cancelled " + result);
		this.listener.onDone(result);
	}

	/**
	 * SizeDownloader listener
	 */
	@FunctionalInterface
	interface Listener
	{
		/**
		 * Done
		 *
		 * @param result file data
		 */
		void onDone(final FileData result);
	}

	/**
	 * Start file data downloading
	 *
	 * @param activity           launching activity
	 * @param name               name of file to be downloaded
	 * @param downloadSourceUrl  download source url
	 * @param downloadIntent     downloader intent (activity launched if update is requested)
	 */
	static public void start(@NonNull final Activity activity, @Nullable final String name, @Nullable final String downloadSourceUrl, final Intent downloadIntent)
	{
		// download source data
		if (name == null || downloadSourceUrl == null || downloadSourceUrl.isEmpty())
		{
			final String message = activity.getString(R.string.status_download_error_unavailable_download_url);
			activity.runOnUiThread(() -> Toast.makeText(activity, message, Toast.LENGTH_LONG).show());
			return;
		}

		// download source data (acquired by task)
		final FileDataDownloader task = new FileDataDownloader(srcData -> {

			// in-use downstream data
			final long downDateValue = Settings.getDatapackDate(activity);
			final long downSizeValue = Settings.getDatapackSize(activity);
			final Date downDate = downDateValue == -1 || downDateValue == 0 ? null : new Date(downDateValue);
			final Long downSize = downSizeValue == -1 ? null : downSizeValue;

			// in-use downstream recorded source data
			final String downSource = Settings.getDatapackSource(activity);
			final long downSourceDateValue = Settings.getDatapackSourceDate(activity);
			final long downSourceSizeValue = Settings.getDatapackSourceSize(activity);
			final Date downSourceDate = downSourceDateValue == -1 || downSourceDateValue == 0 ? null : new Date(downSourceDateValue);
			final Long downSourceSize = downSourceSizeValue == -1 ? null : downSourceSizeValue;
			final String downSourceEtag = Settings.getDatapackSourceEtag(activity);
			final String downSourceVersion = Settings.getDatapackSourceVersion(activity);
			final String downSourceStaticVersion = Settings.getDatapackSourceStaticVersion(activity);

			// upstream data from http connection
			final Date srcDate = srcData == null ? null : srcData.getDate();
			final Long srcSize = srcData == null ? null : srcData.getSize();
			final String srcEtag = srcData == null ? null : srcData.getEtag();
			final String srcVersion = srcData == null ? null : srcData.getVersion();
			final String srcStaticVersion = srcData == null ? null : srcData.getStaticVersion();

			// newer
			final boolean same = (srcEtag != null && srcEtag.equals(downSourceEtag)) || (srcVersion != null && srcVersion.equals(downSourceVersion)) || (srcStaticVersion != null && srcStaticVersion.equals(downSourceStaticVersion)); // one match
			final boolean newer = srcDate == null || downDate == null || srcDate.compareTo(downDate) > 0; // upstream src date from http connection is newer than recorded date
			final boolean srcNewer = srcDate == null || downSourceDate == null || srcDate.compareTo(downSourceDate) > 0; // upstream src date from http connection is newer than recorded source date

			// start update activity
			final Intent intent = new Intent(activity, UpdateActivity.class);

			// result
			intent.putExtra(UpdateFragment.UP_SOURCE_ARG, downloadSourceUrl);
			intent.putExtra(UpdateFragment.UP_DATE_ARG, srcDate == null ? "n/a" : srcDate.toString());
			intent.putExtra(UpdateFragment.UP_SIZE_ARG, srcSize == null ? "n/a" : srcSize + " bytes");
			intent.putExtra(UpdateFragment.UP_ETAG_ARG, srcEtag == null ? "n/a" : srcEtag);
			intent.putExtra(UpdateFragment.UP_VERSION_ARG, srcVersion == null ? "n/a" : srcVersion);
			intent.putExtra(UpdateFragment.UP_STATIC_VERSION_ARG, srcStaticVersion == null ? "n/a" : srcStaticVersion);

			intent.putExtra(UpdateFragment.DOWN_NAME_ARG, name);
			intent.putExtra(UpdateFragment.DOWN_DATE_ARG, downDate == null ? "n/a" : downDate.toString());
			intent.putExtra(UpdateFragment.DOWN_SIZE_ARG, downSize == null ? "n/a" : downSize + " bytes");
			intent.putExtra(UpdateFragment.DOWN_SOURCE_ARG, downSource);
			intent.putExtra(UpdateFragment.DOWN_SOURCE_DATE_ARG, downSourceDate == null ? "n/a" : downSourceDate.toString());
			intent.putExtra(UpdateFragment.DOWN_SOURCE_SIZE_ARG, downSourceSize == null ? "n/a" : downSourceSize + " bytes");
			intent.putExtra(UpdateFragment.DOWN_SOURCE_ETAG_ARG, downSourceEtag == null ? "n/a" : downSourceEtag);
			intent.putExtra(UpdateFragment.DOWN_SOURCE_VERSION_ARG, downSourceVersion == null ? "n/a" : downSourceVersion);
			intent.putExtra(UpdateFragment.DOWN_SOURCE_STATIC_VERSION_ARG, downSourceVersion == null ? "n/a" : downSourceStaticVersion);

			intent.putExtra(UpdateFragment.NEWER_ARG, !same && (newer || srcNewer));

			// to do if confirmed
			intent.putExtra(UpdateFragment.DOWNLOAD_INTENT_ARG, downloadIntent);

			activity.startActivity(intent);
		});
		task.execute(downloadSourceUrl);
	}
}

File: download/src/main/java/com/bbou/download/workers/utils/FileDataDownloader.kt
/*
 * Copyright (c) 2023. Bernard Bou <1313ou@gmail.com>.
 */
package com.bbou.download.workers.utils

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.util.Log
import android.widget.Toast
import com.bbou.concurrency.Task
import com.bbou.download.Keys.DOWNLOAD_FROM_ARG
import com.bbou.download.UpdateActivity
import com.bbou.download.UpdateFragment
import com.bbou.download.common.R
import com.bbou.download.preference.Settings
import com.bbou.download.utils.FileData
import java.net.HttpURLConnection
import java.net.URL
import java.util.Date

/**
 * File data downloader
 *
 * @author [Bernard Bou](mailto:1313ou@gmail.com)
 */
class FileDataDownloader(private val listener: Listener) : Task<String, Void, FileData?>() {

    override fun doJob(params: String?): FileData? {
        val urlString = params!!
        var date: Long = -1
        var size: Long = -1
        var etag: String? = null
        var version: String? = null
        var staticVersion: String? = null
        var httpConnection: HttpURLConnection? = null
        try {
            // url
            val url = URL(urlString)
            Log.d(TAG, "Getting $url")

            // connection
            var connection = url.openConnection()

            // handle redirect
            if (connection is HttpURLConnection) {
                httpConnection = connection
                httpConnection.instanceFollowRedirects = false
                HttpURLConnection.setFollowRedirects(false)
                val status = httpConnection.responseCode
                Log.d(TAG, "Response Code ... $status")
                if (status == HttpURLConnection.HTTP_MOVED_TEMP || status == HttpURLConnection.HTTP_MOVED_PERM || status == HttpURLConnection.HTTP_SEE_OTHER) {
                    // headers
                    date = connection.lastModified // new Date(date))
                    size = connection.contentLength.toLong()
                    etag = connection.getHeaderField("etag")
                    version = connection.getHeaderField("x-version")
                    staticVersion = connection.getHeaderField("x-static-version")

                    // get redirect url from "location" header field
                    val newUrl = httpConnection.getHeaderField("Location")

                    // close
                    httpConnection.inputStream.close()

                    // disconnect
                    httpConnection.disconnect()

                    // open the new connection again
                    httpConnection = URL(newUrl).openConnection() as HttpURLConnection
                    connection = httpConnection
                    httpConnection.instanceFollowRedirects = true
                    HttpURLConnection.setFollowRedirects(true)
                    Log.d(TAG, "Redirect to URL : $newUrl")
                }
            }

            // connect
            connection.connect()

            // expect HTTP 200 OK, so we don't mistakenly save error report instead of the file
            if (connection is HttpURLConnection) {
                if (httpConnection!!.responseCode != HttpURLConnection.HTTP_OK) {
                    val message = "server returned HTTP " + httpConnection.responseCode + " " + httpConnection.responseMessage
                    throw RuntimeException(message)
                }
            }
            val name = url.file
            if (date <= 0) {
                date = connection.lastModified // new Date(date))
            }
            if (size <= 0) {
                size = connection.contentLength.toLong()
            }
            if (etag == null) {
                etag = connection.getHeaderField("etag")
            }
            if (version == null) {
                version = connection.getHeaderField("x-version")
            }
            if (staticVersion == null) {
                staticVersion = connection.getHeaderField("x-static-version")
            }

            // close
            connection.getInputStream().close()
            return FileData(name, date, size, etag, version, staticVersion)
        } catch (e: Exception) {
            Log.e(TAG, "While downloading", e)
        } finally {
            httpConnection?.disconnect()
        }
        return null
    }

    override fun onDone(result: FileData?) {
        Log.d(TAG, "Completed $result")
        listener.onDone(result)
    }

    override fun onCancelled() {
        Log.d(TAG, "Cancelled")
        listener.onDone(null)
    }

    /**
     * File data downloader listener
     */
    fun interface Listener {

        /**
         * Done
         *
         * @param result file data
         */
        fun onDone(result: FileData?)
    }

    companion object {

        private const val TAG = "FileDataDownloader"

        /**
         * Start file data downloading
         *
         * @param activity       launching activity
         * @param downloadIntent download intent (activity launched if update is requested)
         */
        fun start(activity: Activity, downloadIntent: Intent) {
            // unmarshal arguments
            val downloadSourceUrl = downloadIntent.getStringExtra(DOWNLOAD_FROM_ARG)
            val name = Uri.parse(downloadSourceUrl).lastPathSegment

            // download source data
            if (downloadSourceUrl.isNullOrEmpty() || name == null) {
                val message = activity.getString(R.string.no_url)
                activity.runOnUiThread { Toast.makeText(activity, message, Toast.LENGTH_LONG).show() }
                return
            }

            // download source data (acquired by task)
            val task = FileDataDownloader { srcData: FileData? ->

                // in-use downstream data
                val downDateValue = Settings.getDatapackDate(activity)
                val downSizeValue = Settings.getDatapackSize(activity)
                val downDate = if (downDateValue == -1L || downDateValue == 0L) null else Date(downDateValue)
                val downSize = if (downSizeValue == -1L) null else downSizeValue

                // in-use downstream recorded source data
                val downSource = Settings.getDatapackSource(activity)
                val downSourceDateValue = Settings.getDatapackSourceDate(activity)
                val downSourceSizeValue = Settings.getDatapackSourceSize(activity)
                val downSourceDate = if (downSourceDateValue == -1L || downSourceDateValue == 0L) null else Date(downSourceDateValue)
                val downSourceSize = if (downSourceSizeValue == -1L) null else downSourceSizeValue
                val downSourceEtag = Settings.getDatapackSourceEtag(activity)
                val downSourceVersion = Settings.getDatapackSourceVersion(activity)
                val downSourceStaticVersion = Settings.getDatapackSourceStaticVersion(activity)

                // upstream data from http connection
                val srcDate = srcData?.getDate()
                val srcSize = srcData?.getSize()
                val srcEtag = srcData?.etag
                val srcVersion = srcData?.version
                val srcStaticVersion = srcData?.staticVersion

                // newer
                val same = srcEtag != null && srcEtag == downSourceEtag || srcVersion != null && srcVersion == downSourceVersion || srcStaticVersion != null && srcStaticVersion == downSourceStaticVersion // one match
                val newer = srcDate == null || downDate == null || srcDate > downDate // upstream src date from http connection is newer than recorded date
                val srcNewer = srcDate == null || downSourceDate == null || srcDate > downSourceDate // upstream src date from http connection is newer than recorded source date

                // start update activity
                val intent = Intent(activity, UpdateActivity::class.java)

                // result
                val na = activity.getString(R.string.na)
                val bytes = activity.getString(R.string.bytes)
                intent.putExtra(UpdateFragment.UP_SOURCE_ARG, downloadSourceUrl)
                intent.putExtra(UpdateFragment.UP_DATE_ARG, srcDate?.toString() ?: na)
                intent.putExtra(UpdateFragment.UP_SIZE_ARG, if (srcSize == null) na else "$srcSize $bytes")
                intent.putExtra(UpdateFragment.UP_ETAG_ARG, srcEtag ?: na)
                intent.putExtra(UpdateFragment.UP_VERSION_ARG, srcVersion ?: na)
                intent.putExtra(UpdateFragment.UP_STATIC_VERSION_ARG, srcStaticVersion ?: na)
                intent.putExtra(UpdateFragment.DOWN_NAME_ARG, name)
                intent.putExtra(UpdateFragment.DOWN_DATE_ARG, downDate?.toString() ?: na)
                intent.putExtra(UpdateFragment.DOWN_SIZE_ARG, if (downSize == null) na else "$downSize $bytes")
                intent.putExtra(UpdateFragment.DOWN_SOURCE_ARG, downSource ?: activity.getString(R.string.unsourced))
                intent.putExtra(UpdateFragment.DOWN_SOURCE_DATE_ARG, downSourceDate?.toString() ?: na)
                intent.putExtra(UpdateFragment.DOWN_SOURCE_SIZE_ARG, if (downSourceSize == null) na else "$downSourceSize $bytes")
                intent.putExtra(UpdateFragment.DOWN_SOURCE_ETAG_ARG, downSourceEtag ?: na)
                intent.putExtra(UpdateFragment.DOWN_SOURCE_VERSION_ARG, downSourceVersion ?: na)
                intent.putExtra(UpdateFragment.DOWN_SOURCE_STATIC_VERSION_ARG, if (downSourceVersion == null) na else downSourceStaticVersion)
                intent.putExtra(UpdateFragment.NEWER_ARG, !same && (newer || srcNewer))

                // to do if confirmed
                intent.putExtra(UpdateFragment.DOWNLOAD_INTENT_ARG, downloadIntent)
                activity.startActivity(intent)
            }
            task.execute(downloadSourceUrl)
        }
    }
}
